### ISSUE TITLE
keyboard_enter_text error "Timed out waiting for key event to complete"

### ISSUE THREAD

**Issue by bhaskarsr** [calabash/calabash-ios#1249]
Hi There,

keyboard_enter_text most of the times throw the error 

"Expected JSON response with no error, but found
Timed out waiting for key event to complete
(RunLoop::DeviceAgent::Client::HTTPError)" 

It actually enters one single character in the text field and then fails.

When I run the same scenario, it works fine. When I execute my complete test suite with a lot, the volume of the scenarios that fail with this error is even more. Specific screens in our app fail more frequently with this error.
I am facing this issue since I updated my XCode to 8.1. Not sure if any one has faced this issue or not.

Can you please help me out here if I am missing something?

Thanks in advance.

**Comment by bhaskarsr** [comment]
I am using Calabash-IOS version 0.20.3.

**Comment by bhaskarsr** [comment]
Hi @jmoody, Could you please assist here if anything is missing ?

**Comment by nandhukishore04** [comment]
I am also facing this issue from today. gem using calabash-cucumber (= 0.20.0)

**Comment by ricmatsui** [comment]
I am also seeing this issue, here are some more details. It seems like there is a timeout being triggered in `testmanagerd` possibly due to a delay in typing the character.

iOS 10.2 Simulator iPhone 7
Xcode 8.2
calabash-cucumber 0.20.3
calabash.framework 0.20.3

Example Step Definition
```
    wait_for_keyboard
    keyboard_enter_text 'calabash@example.com'
```

Example Error
```
      
      Expected JSON response with no error, but found
      
      Timed out waiting for key event to complete
      
       (RunLoop::DeviceAgent::Client::HTTPError)
```

Console
```
testmanagerd	10:23:08.635 testmanagerd[56738:380699] Typing calabash@example.com
testmanagerd	10:23:08.635 testmanagerd[56738:380699] Typing calabash@example.com
testmanagerd	10:23:08.636 testmanagerd[56738:380703] Key event 'c' sent.
testmanagerd	10:23:08.636 testmanagerd[56738:380703] Key event 'c' sent.
testmanagerd	10:23:09.708 testmanagerd[56738:380699] Timed out after waiting 1.0s for KeyEventCompleted after sending event for 'c'.
testmanagerd	10:23:09.708 testmanagerd[56738:380699] Timed out after waiting 1.0s for KeyEventCompleted after sending event for 'c'.
testmanagerd	10:23:09.709 testmanagerd[56738:380699] Typing complete, replying to client.
testmanagerd	10:23:09.709 testmanagerd[56738:380699] Typing complete, replying to client.
```

**Comment by jmoody** [comment]
We are aware that this error can happen regularly on 32 bit devices (i386 simulators and armv7* devices).  On 64-bit devices, we think we have solved this problem in run_loop 2.2.4.  If you are using run-loop 2.2.4 and you are still seeing this problem, we want to know about it.

**Comment by bhaskarsr** [comment]
@jmoody I am using run-loop 2.2.4 version and have this problem.

**Comment by ricmatsui** [comment]
Thanks @jmoody, I will let you know if we see the issue with run_loop 2.2.4, I was using 2.2.3.

**Comment by nandhukishore04** [comment]
@jmoody I am using iphone 5s which is 64 bit and run_loop version 2.5.1. still facing this issue. If one scenario starts failing by this error then all the upcoming scenarios will get fail with same. Kindly help on this. Thanks

**Comment by jmoody** [comment]
@nandhukishore04 Can you confirm your run-loop version?  2.5.1 does not exist.

This issue is at the top of my queue and we are actively working on a solution.

**Comment by nandhukishore04** [comment]
@jmoody sorry its my mistake. I tried with the command gem -v run_loop and it showed 2.5.1. This is my content in gemlock file  run_loop (>= 2.2.0, < 3.0)

**Comment by nandhukishore04** [comment]
I think this issue is only in device agent iOS10. After changing the iOS version to 9, it is working fine.

**Comment by tommeier** [comment]
Seeing this, rarely, against iOS 10.2 on Xcode 8.2, via calabash 0.20.4 + run_loop 2.2.4. Wrapping my keyboard entry commands and retrying a max of 3 times on any `RunLoop::DeviceAgent::Client::HTTPError` errors on entry.

**Comment by jmoody** [comment]
@tommeier An interesting solution.  We know that once this error occurs, the DeviceAgent and/or device is in a bad state for testing - subsequent gestures will fail.

**Comment by mriddle** [comment]
@jmoody I started experiencing this issue too but only after I'd updated [run_loop to include the latest changes](https://github.com/calabash/run_loop/compare/v2.2.4...d9099cb247f6105c0efbbc887fb82e87226c5475) (resolves the [slow iOS 10 issues](https://github.com/calabash/calabash-ios/issues/1234)). Scenarios run in isolation would work but when running a feature file only the first scenario would work. The rest would fail with the error described above.

I was able to resolve it by [stopping the agent instead of calling session delete](https://github.com/calabash/calabash-ios/compare/develop...mriddle:session-shutdown#diff-702004ee813f28d7c7d9040bde9c765dL1532). 

However this introduced a new problem whereby any features using the `RESET_BETWEEN_SCENARIOS` environment variables would fail with "`Expected 'Shutdown' but found 'Booted' after waiting.`". I think this was caused by the refactor you did in run_loop and I was able to resolve it by [quitting the simulator when resetting the app sandbox](https://github.com/calabash/run_loop/compare/develop...mriddle:quit-on-reset#diff-922955b0b6721c96dda90391ea96ed9cR436)

With the changes described above I was able to get our builds working again and running at the roughly the same speed before the move from the iOS 9 to iOS 10 simulators.

versions: iOS 10.2, xcode 8.2, calabash 0.20.4 (with change above), run_loop @ `d9099cb` (with change avove)

**Comment by nadzeya** [comment]
I have that issue as well on iPhone 6s iOS 10.2, Xcode 8.2, calabash 0.20.4, run_loop 2.2.4

**Comment by jaysacco** [comment]
I have just run into this problem with calabash-ios 0.20.4/Xcode 8.2.1/run_loop 2.2.4 testing on an iPhone 6s Plus simulator running iOS 9.3.  I can also provide other info if needed.

**Comment by jmoody** [comment]
We are aware of this problem and continue to make improvements.  run-loop 2.3.0 will provide some relief, but not a complete fix.

**Comment by sowmyahebbar** [comment]
Would like to know when will run-loop 2.3.0 will be available.

**Comment by lararufflecoles** [comment]
I also need to know when this update will be available.

**Comment by bengoism** [comment]
I got around this problem by disabling the "Connect Hardware keyboard" option in the simulator menu

Hardware > Keyboard >  Connect Hardware keyboard

**Comment by jmoody** [comment]
These errors have been resolved.  Please open a new issue if they happen again.

**Comment by AartiJ** [comment]
which run_loop version has fix for this?

**Comment by suchikaashyap** [comment]
@tommeier we are encountering this issue even after the calabash upgrade to 0.20.5 and run_loop 2.4.1.
I have tried the handling of this error within the step but as @jmoody insinuated it doesn't seem to aleviate the problem. Do you have any sample code that you can share perhaps?

**Comment by tommeier** [comment]
@suchikaashyap Hi! After upgrading ios simulator it seems to be far more stable now. So the retry isn't even being triggered right now.

**Comment by Rain2333333** [comment]
yes, i know this problem. 
when it's happend ,u should setting simulator  just like what @bengoism  say.
Then if ur simulator waiting for something
at this moment u could do something like this
Hardware > Keyboard >Toggle Software Kerboard

**Comment by otamas94** [comment]
And if you use physical device instead of a simulator? 

I have on iOS 11.2.0 and an iOS 10.3.2 on separated VMs, with the exactly same setup:
ruby 2.0.0
calabash-ios 0.21.5
run_loop 2.7.0

After 2 hours of running on iOS 10.3.2 every test start to fail with:
`Expected JSON response with no error, but found
13:59:26       
13:59:26       Timed out waiting for key event to complete
13:59:26       
13:59:26        (RunLoop::DeviceAgent::Client::HTTPError)`

It types in the 1st character of the text then fails, but it's not a problem with the length, because if I type in 1 char wth keyboard_enter_text it fails again.

On iOS 11 same test set works completely fine, on iOS10 the issue occurs every time when I run my test set for more then 2 hours.
